Update to latest riak_ensemble integrity approach

jtuple opened the pull request:
This is the sibling pull-request for basho/riak_ensemble#37. That pull-request completely changes how `riak_ensemble` performs data validation and integrity checking, completely removing the need for backend based trust/syncing.

This pull-request updates `riak_kv` to work with this new approach. For the most part, that means deleting a bunch of code that is no longer necessary. As well as updating the `ensemble_status` command to no longer reference AAE or trust -- since neither of those concepts apply anymore.

For more details, look at the individual commits.

**_Note:**_ a discussion with product management needs to happen concerning the new output of `ensemble_status` as well as how we should expose the new integrity checking settings to `riak.conf`. The old discussion about `trust` no longer applies since that setting has been removed. I think the approach as implemented in this pull-request is good enough for 2.0 RC But, we need to have that discussion / change things before 2.0 final.

jtuple committed:
Remove ensemble backend syncing logic

riak_ensemble now supports an integrated integrity checking mechanism,
and therefore no longer requires nor supports backend based trust and
syncing logic. This commit removes all such logic from riak_kv.

jtuple committed:
Update consistent AAE to simply use read repair

The previous approach for consistent AAE was designed to function as the
backend syncing mechanism for K/V ensembles. Now that riak_ensemble no
longer requires backend syncing, this approach is no longer necessary
nor is it correct.

This commit completely removes the previous approach, reducing consistent
AAE to simple read repair now that riak_ensemble has a read repair mechanism.
This not only simplifies things, but this approach also properly works with
the new integrated integrity/validation logic in riak_ensemble.

jtuple committed:
Add synctree_path callback to ensemble backend

This commit adds the new synctree_path callback to the K/V ensemble
backend. This callback is used to override how riak_ensemble maps
peer synctrees to on-disk LevelDB instances.

This commit implements the callback using a M:N mapping, where all
peers that correspond to a single vnode map to a common LevelDB
instance. This makes the number of synctree LevelDB instances scale
with the size of the ring, rather than the number of distinct preflists.

This is the same approach used for AAE in hashtree.erl.

Requiring fewer instances makes better use of file handles, write buffers,
and internal LevelDB caches.

jtuple committed:
Remove final remnants of ensemble backend syncing

Update riak_kv_ensemble_console to no longer talk about trust or AAE,
as backend based trust and AAE syncing has been completely removed.

The console output now references concepts relevant to the new
synctree approach directly integrated into riak_ensemble.

Remove the trust setting from riak_kv.schema since the setting no
longer exists. Providing schema mappings for the new relevant settings
is left for a future commit.

jtuple committed:
Update spec for riak_kv_ensemble_console:ping_quorum/1

jtuple commented:
Concerning the builder failure, this pull-request will fail until the sibling `riak_ensemble` pull-request is merged.

engelsanchez commented on src/riak_kv_ensemble_console.erl:
```diff
@@ -386,17 +399,11 @@ print_variations() ->
 
-    %% Enabled, enough nodes, but AAE not enabled
-    io:format("(*** AAE disabled when required ***)~n~n"),
-    Details3 = Details2#details{nodes=['dev1@127.0.0.1',
```
Poor Details3 got lost in the shuffle.

jtuple replied:
Aye, easier to just kill it then renumber all the other lines. Keeping diff minimal/etc.

engelsanchez commented on src/riak_kv_exchange_fsm.erl:
```diff
@@ -304,31 +296,11 @@ read_repair_keydiff(RC, LocalVN, RemoteVN, {Bucket, Key, Reason}) ->
-    end.
+repair_consistent(BKey) ->
+    Ensemble = riak_client:ensemble(BKey),
+    Timeout = 60000,
```
Maybe just byte the bullet and make this configurable in advanced config?

jtuple replied:
We could, but maybe later? Normal Riak read repair triggered by AAE is just a `riak_client:get` which uses a `60000` hardcoded timeout in `riak_client`. Figured copying existing behavior was good enough for now.

engelsanchez replied:
Yes, we can do config cleanup later.

andrewjstone commented:
:+1: 1519686
Code looks good and riak tests pass with the following changes:
- https://github.com/basho/riak_test/pull/657
- https://github.com/basho/riak_test/pull/658
- https://github.com/basho/riak_ensemble/pull/37

jtuple committed:
Update riak_kv_schema_tests

The riak_kv.consistency_trust setting has been removed, therefore this
commit updates riak_kv_schema_tests to no longer test for this setting.

jtuple commented:
@borshop merge
